File: backend-java/src/main/java/com/onlineshop/service/AdminStatsService.java
package com.onlineshop.service;

import com.onlineshop.dto.AdminDtos.*;
import com.onlineshop.dto.TemplateDto;
import com.onlineshop.repository.RoomTemplateRepository;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;

/**
 * Java port of internal/service/AdminStatsService.
 *
 * Aggregations use native SQL since the underlying queries are relatively
 * complex; this matches the sqlc-generated query layer in Go.
 */
@Service
@RequiredArgsConstructor
public class AdminStatsService {

    private final RoomTemplateRepository templateRepo;

    @PersistenceContext
    private EntityManager em;

    @Transactional(readOnly = true)
    public HistoricalMetrics getHistoricalMetrics() {
        Instant since = Instant.now().minus(7, ChronoUnit.DAYS);
        // Mirrors Go GetHistoricalMetrics in repository/admin.sql.go: real-player count
        // EXCLUDES bots via JOIN users + FILTER (WHERE u.bot = false).
        Object[] row = (Object[]) em.createNativeQuery("""
                WITH completed_rooms AS (
                    SELECT
                        r.room_id,
                        r.entry_cost,
                        COUNT(DISTINCT rp.user_id) FILTER (WHERE u.bot = false) AS real_players
                    FROM rooms r
                    LEFT JOIN room_players rp ON r.room_id = rp.room_id
                    LEFT JOIN users u ON rp.user_id = u.id
                    WHERE r.status = 'finished'
                      AND r.created_at >= :since
                    GROUP BY r.room_id, r.entry_cost
                )
                SELECT
                    COALESCE(AVG(real_players), 0)::float8 AS avg_real_players_per_room,
                    COALESCE(AVG(entry_cost), 0)::float8   AS avg_entry_cost,
                    COUNT(*)::bigint                       AS total_completed_rooms
                FROM completed_rooms
                """)
                .setParameter("since", since)
                .getSingleResult();

        return new HistoricalMetrics(
                ((Number) row[0]).doubleValue(),
                ((Number) row[1]).doubleValue(),
                ((Number) row[2]).longValue());
    }

    @Transactional(readOnly = true)
    public List<Warning> validateTemplate(TemplateDto dto) {
        return validateWithMetrics(dto, getHistoricalMetrics());
    }

    /** Pure function — extracted so it is unit-testable without a database. */
    public static List<Warning> validateWithMetrics(TemplateDto dto, HistoricalMetrics m) {
        List<Warning> w = new ArrayList<>();

        if (dto.playersNeeded() == 1)
            w.add(new Warning("players_needed", "Player count is too low (1 player)", "warning"));
        if (m.avgRealPlayersPerRoom() > 0 && dto.minPlayers() > m.avgRealPlayersPerRoom())
            w.add(new Warning("min_players", "Minimum players exceeds average real player count per room from past week", "warning"));
        if (m.avgRealPlayersPerRoom() > 0 && dto.playersNeeded() > 2 * m.avgRealPlayersPerRoom())
            w.add(new Warning("players_needed", "Maximum players exceeds twice the average real player count per room from past week", "warning"));
        if (m.avgEntryCost() > 0 && dto.entryCost() > 1.75 * m.avgEntryCost())
            w.add(new Warning("entry_cost", "Entry cost is too high (exceeds 1.75x average)", "warning"));
        if (m.avgEntryCost() > 0 && dto.entryCost() < 0.5 * m.avgEntryCost())
            w.add(new Warning("entry_cost", "Entry cost is too low (less than 0.5x average)", "warning"));
        if (dto.winnerPct() > 80)
            w.add(new Warning("winner_pct", "Jackpot percentage is too high (exceeds 80%)", "warning"));
        if (dto.winnerPct() < 50)
            w.add(new Warning("winner_pct", "Jackpot percentage is too low (less than 50%)", "warning"));
        return w;
    }

    @Transactional(readOnly = true)
    public boolean checkDuplicate(TemplateDto dto) {
        return templateRepo.existsDuplicate(
                dto.maxPlayers(), dto.minPlayers(), dto.entryCost(), dto.winnerPct(),
                dto.effectiveGameType());
    }

    /**
     * Mirrors Go {@code AdminStatsService.ValidateTemplate}:
     * uses only the 5 fields available on the admin payload (no name/round/delay).
     */
    @Transactional(readOnly = true)
    public List<Warning> validateAdminTemplate(AdminValidateTemplateRequest dto) {
        HistoricalMetrics m = getHistoricalMetrics();
        List<Warning> w = new ArrayList<>();

        if (dto.playersNeeded() == 1)
            w.add(new Warning("players_needed", "Player count is too low (1 player)", "warning"));
        if (m.avgRealPlayersPerRoom() > 0 && dto.minPlayers() > m.avgRealPlayersPerRoom())
            w.add(new Warning("min_players", "Minimum players exceeds average real player count per room from past week", "warning"));
        if (m.avgRealPlayersPerRoom() > 0 && dto.playersNeeded() > 2 * m.avgRealPlayersPerRoom())
            w.add(new Warning("players_needed", "Maximum players exceeds twice the average real player count per room from past week", "warning"));
        if (m.avgEntryCost() > 0 && dto.entryCost() > 1.75 * m.avgEntryCost())
            w.add(new Warning("entry_cost", "Entry cost is too high (exceeds 1.75x average)", "warning"));
        if (m.avgEntryCost() > 0 && dto.entryCost() < 0.5 * m.avgEntryCost())
            w.add(new Warning("entry_cost", "Entry cost is too low (less than 0.5x average)", "warning"));
        if (dto.winnerPct() > 80)
            w.add(new Warning("winner_pct", "Jackpot percentage is too high (exceeds 80%)", "warning"));
        if (dto.winnerPct() < 50)
            w.add(new Warning("winner_pct", "Jackpot percentage is too low (less than 50%)", "warning"));
        return w;
    }

    @Transactional(readOnly = true)
    public boolean checkDuplicateAdmin(AdminValidateTemplateRequest dto) {
        com.onlineshop.domain.enums.GameType gt = dto.gameType() == null || dto.gameType().isBlank()
                ? com.onlineshop.domain.enums.GameType.FRIDGE
                : com.onlineshop.domain.enums.GameType.fromValue(dto.gameType());
        return templateRepo.existsDuplicate(
                dto.playersNeeded(), dto.minPlayers(), dto.entryCost(), dto.winnerPct(), gt);
    }

    @Transactional(readOnly = true)
    public TemplateStats getTemplateStatistics(int templateId, TimeFilter filter) {
        Instant[] range = parseRange(filter);
        Instant start = range[0], end = range[1];

        // Mirrors Go GetTemplateStatistics: includes users JOIN with bot=false / bot=true filters.
        Object[] tpl = (Object[]) em.createNativeQuery("""
                WITH template_rooms AS (
                    SELECT r.room_id
                    FROM rooms r
                    WHERE r.template_id = :tid
                      AND r.status = 'finished'
                      AND r.created_at >= :start
                      AND r.created_at <= :end
                )
                SELECT
                    COUNT(DISTINCT tr.room_id)::int AS completed_rooms,
                    COUNT(DISTINCT rp.user_id) FILTER (WHERE u.bot = false)::int AS total_real_players,
                    COUNT(DISTINCT rp.user_id) FILTER (WHERE u.bot = true)::int  AS total_bots,
                    COALESCE(
                        COUNT(DISTINCT rp.user_id) FILTER (WHERE u.bot = false)::float8 /
                        NULLIF(COUNT(DISTINCT tr.room_id), 0),
                        0
                    )::float8 AS avg_real_players_per_room
                FROM template_rooms tr
                LEFT JOIN room_players rp ON tr.room_id = rp.room_id
                LEFT JOIN users u ON rp.user_id = u.id
                """)
                .setParameter("tid", templateId)
                .setParameter("start", start)
                .setParameter("end", end)
                .getSingleResult();

        // Mirrors Go GetWinnerStatistics: split real/bot wins via users.bot.
        Object[] win = (Object[]) em.createNativeQuery("""
                SELECT
                    COUNT(*) FILTER (WHERE u.bot = false)::int AS real_player_wins,
                    COUNT(*) FILTER (WHERE u.bot = true)::int  AS bot_wins
                FROM room_winners rw
                INNER JOIN users u ON rw.user_id = u.id
                INNER JOIN rooms r ON r.room_id = rw.room_id
                WHERE r.template_id = :tid
                  AND rw.won_at >= :start AND rw.won_at <= :end
                """)
                .setParameter("tid", templateId)
                .setParameter("start", start)
                .setParameter("end", end)
                .getSingleResult();

        // Mirrors Go GetBoostStatistics: per-player avg = SUM/COUNT(DISTINCT user), not AVG(amount).
        Object[] boost = (Object[]) em.createNativeQuery("""
                WITH boost_data AS (
                    SELECT
                        SUM(rb.amount)::bigint AS total_boost_amount,
                        COUNT(DISTINCT rb.user_id)::int AS total_players,
                        COUNT(DISTINCT rb.room_id)::int AS total_rooms
                    FROM room_boosts rb
                    INNER JOIN rooms r ON rb.room_id = r.room_id
                    WHERE r.template_id = :tid
                      AND rb.boosted_at >= :start AND rb.boosted_at <= :end
                )
                SELECT
                    COALESCE(total_boost_amount, 0)::bigint AS total_boost_amount,
                    COALESCE(total_boost_amount::float8 / NULLIF(total_players, 0), 0)::float8 AS avg_boost_per_player,
                    COALESCE(total_boost_amount::float8 / NULLIF(total_rooms, 0), 0)::float8   AS avg_boost_per_room
                FROM boost_data
                """)
                .setParameter("tid", templateId)
                .setParameter("start", start)
                .setParameter("end", end)
                .getSingleResult();

        // Mirrors Go GetPlaceStatistics: total_places / DISTINCT players (not avg per group).
        Number avgPlaces = (Number) em.createNativeQuery("""
                WITH place_data AS (
                    SELECT
                        COUNT(*)::int AS total_places,
                        COUNT(DISTINCT rpl.user_id)::int AS total_players
                    FROM room_places rpl
                    INNER JOIN rooms r ON rpl.room_id = r.room_id
                    WHERE r.template_id = :tid
                      AND rpl.created_at >= :start AND rpl.created_at <= :end
                )
                SELECT COALESCE(total_places::float8 / NULLIF(total_players, 0), 0)::float8
                FROM place_data
                """)
                .setParameter("tid", templateId)
                .setParameter("start", start)
                .setParameter("end", end)
                .getSingleResult();

        return new TemplateStats(
                templateId,
                ((Number) tpl[0]).intValue(),
                ((Number) tpl[1]).intValue(),
                ((Number) tpl[2]).intValue(),
                ((Number) tpl[3]).doubleValue(),
                ((Number) win[0]).intValue(),
                ((Number) win[1]).intValue(),
                ((Number) boost[0]).longValue(),
                ((Number) boost[1]).doubleValue(),
                ((Number) boost[2]).doubleValue(),
                avgPlaces.doubleValue());
    }

    public static Instant[] parseRange(TimeFilter f) {
        Instant now = Instant.now();
        Instant start, end;
        String period = f == null || f.period() == null ? "all" : f.period();
        switch (period) {
            case "hour"  -> { start = now.minus(1, ChronoUnit.HOURS); end = now; }
            case "day"   -> { start = now.minus(1, ChronoUnit.DAYS); end = now; }
            case "week"  -> { start = now.minus(7, ChronoUnit.DAYS); end = now; }
            case "month" -> { start = now.minus(30, ChronoUnit.DAYS); end = now; }
            case "custom" -> {
                start = f.startTime() != null ? f.startTime() : Instant.parse("2000-01-01T00:00:00Z");
                end = f.endTime() != null ? f.endTime() : now;
            }
            default -> { start = Instant.parse("2000-01-01T00:00:00Z"); end = now; }
        }
        return new Instant[]{start, end};
    }

    /** Convenience: BigDecimal helper for numeric coercion. */
    @SuppressWarnings("unused")
    private static double toDouble(Object o) {
        if (o instanceof BigDecimal b) return b.doubleValue();
        if (o instanceof Number n) return n.doubleValue();
        return 0.0;
    }
}

File: backend/internal/service/admin_stats_service.go
package service

import (
	"context"
	"time"

	"github.com/SomeSuperCoder/OnlineShop/repository"
	"github.com/jackc/pgx/v5/pgxpool"
)

// AdminStatsService provides business logic for admin statistics and validation
type AdminStatsService struct {
	repo *repository.Queries
	pool *pgxpool.Pool
}

// NewAdminStatsService creates a new AdminStatsService instance
func NewAdminStatsService(repo *repository.Queries, pool *pgxpool.Pool) *AdminStatsService {
	return &AdminStatsService{
		repo: repo,
		pool: pool,
	}
}

// HistoricalMetrics contains aggregated metrics from past gameplay
type HistoricalMetrics struct {
	AvgRealPlayersPerRoom float64
	AvgEntryCost          float64
	TotalRooms            int64
}

// TemplateStats contains comprehensive statistics for a template
type TemplateStats struct {
	TemplateID            int32   `json:"template_id"`
	CompletedRooms        int32   `json:"completed_rooms"`
	TotalRealPlayers      int32   `json:"total_real_players"`
	TotalBots             int32   `json:"total_bots"`
	AvgRealPlayersPerRoom float64 `json:"avg_real_players_per_room"`
	RealPlayerWins        int32   `json:"real_player_wins"`
	BotWins               int32   `json:"bot_wins"`
	TotalBoostAmount      int64   `json:"total_boost_amount"`
	AvgBoostPerPlayer     float64 `json:"avg_boost_per_player"`
	AvgBoostPerRoom       float64 `json:"avg_boost_per_room"`
	AvgPlacesPerPlayer    float64 `json:"avg_places_per_player"`
}

// TimeFilter specifies a time period for filtering statistics
type TimeFilter struct {
	Period    string     `json:"period"` // "hour", "day", "week", "month", "all", "custom"
	StartTime *time.Time `json:"start_time,omitempty"`
	EndTime   *time.Time `json:"end_time,omitempty"`
}

// Warning represents a validation warning or error
type Warning struct {
	Field    string `json:"field"`
	Message  string `json:"message"`
	Severity string `json:"severity"` // "warning" or "error"
}

// ValidateTemplateParams contains parameters for template validation
type ValidateTemplateParams struct {
	PlayersNeeded int32
	MinPlayers    int32
	EntryCost     int32
	WinnerPct     int32
	GameType      string
}

// GetHistoricalMetrics retrieves aggregated metrics from the past 7 days
func (s *AdminStatsService) GetHistoricalMetrics(ctx context.Context) (*HistoricalMetrics, error) {
	// Calculate time 7 days ago
	since := time.Now().AddDate(0, 0, -7)

	result, err := s.repo.GetHistoricalMetrics(ctx, repository.GetHistoricalMetricsParams{
		CreatedAt: since,
	})
	if err != nil {
		return nil, err
	}

	return &HistoricalMetrics{
		AvgRealPlayersPerRoom: result.AvgRealPlayersPerRoom,
		AvgEntryCost:          result.AvgEntryCost,
		TotalRooms:            result.TotalCompletedRooms,
	}, nil
}

// ValidateTemplate checks template parameters against historical data and returns warnings
func (s *AdminStatsService) ValidateTemplate(ctx context.Context, params ValidateTemplateParams) ([]Warning, error) {
	var warnings []Warning

	// Get historical metrics for comparison
	metrics, err := s.GetHistoricalMetrics(ctx)
	if err != nil {
		return nil, err
	}

	// Check max players = 1
	if params.PlayersNeeded == 1 {
		warnings = append(warnings, Warning{
			Field:    "players_needed",
			Message:  "Player count is too low (1 player)",
			Severity: "warning",
		})
	}

	// Check min players against historical average
	if metrics.AvgRealPlayersPerRoom > 0 && float64(params.MinPlayers) > metrics.AvgRealPlayersPerRoom {
		warnings = append(warnings, Warning{
			Field:    "min_players",
			Message:  "Minimum players exceeds average real player count per room from past week",
			Severity: "warning",
		})
	}

	// Check max players against historical average (2x threshold)
	if metrics.AvgRealPlayersPerRoom > 0 && float64(params.PlayersNeeded) > 2*metrics.AvgRealPlayersPerRoom {
		warnings = append(warnings, Warning{
			Field:    "players_needed",
			Message:  "Maximum players exceeds twice the average real player count per room from past week",
			Severity: "warning",
		})
	}

	// Check entry cost against historical average (1.75x threshold)
	if metrics.AvgEntryCost > 0 && float64(params.EntryCost) > 1.75*metrics.AvgEntryCost {
		warnings = append(warnings, Warning{
			Field:    "entry_cost",
			Message:  "Entry cost is too high (exceeds 1.75x average)",
			Severity: "warning",
		})
	}

	// Check entry cost against historical average (0.5x threshold)
	if metrics.AvgEntryCost > 0 && float64(params.EntryCost) < 0.5*metrics.AvgEntryCost {
		warnings = append(warnings, Warning{
			Field:    "entry_cost",
			Message:  "Entry cost is too low (less than 0.5x average)",
			Severity: "warning",
		})
	}

	// Check winner_pct > 80
	if params.WinnerPct > 80 {
		warnings = append(warnings, Warning{
			Field:    "winner_pct",
			Message:  "Jackpot percentage is too high (exceeds 80%)",
			Severity: "warning",
		})
	}

	// Check winner_pct < 50
	if params.WinnerPct < 50 {
		warnings = append(warnings, Warning{
			Field:    "winner_pct",
			Message:  "Jackpot percentage is too low (less than 50%)",
			Severity: "warning",
		})
	}

	return warnings, nil
}

// CheckDuplicateTemplate checks if a template with identical parameters already exists
func (s *AdminStatsService) CheckDuplicateTemplate(ctx context.Context, params ValidateTemplateParams) (bool, error) {
	exists, err := s.repo.CheckDuplicateTemplate(ctx, repository.CheckDuplicateTemplateParams{
		PlayersNeeded: params.PlayersNeeded,
		MinPlayers:    params.MinPlayers,
		EntryCost:     params.EntryCost,
		WinnerPct:     params.WinnerPct,
		GameType:      params.GameType,
	})
	if err != nil {
		return false, err
	}

	return exists, nil
}

// GetTemplateStatistics retrieves comprehensive statistics for a template with time filtering
func (s *AdminStatsService) GetTemplateStatistics(ctx context.Context, templateID int32, filter TimeFilter) (*TemplateStats, error) {
	// Parse time filter to get start and end timestamps
	startTime, endTime := parseTimeFilter(filter)

	// Get basic template statistics
	templateStats, err := s.repo.GetTemplateStatistics(ctx, repository.GetTemplateStatisticsParams{
		TemplateID:  &templateID,
		CreatedAt:   startTime,
		CreatedAt_2: endTime,
	})
	if err != nil {
		return nil, err
	}

	// Get winner statistics
	winnerStats, err := s.repo.GetWinnerStatistics(ctx, repository.GetWinnerStatisticsParams{
		TemplateID: &templateID,
		WonAt:      startTime,
		WonAt_2:    endTime,
	})
	if err != nil {
		return nil, err
	}

	// Get boost statistics
	boostStats, err := s.repo.GetBoostStatistics(ctx, repository.GetBoostStatisticsParams{
		TemplateID:  &templateID,
		BoostedAt:   startTime,
		BoostedAt_2: endTime,
	})
	if err != nil {
		return nil, err
	}

	// Get place statistics
	avgPlaces, err := s.repo.GetPlaceStatistics(ctx, repository.GetPlaceStatisticsParams{
		TemplateID:  &templateID,
		CreatedAt:   startTime,
		CreatedAt_2: endTime,
	})
	if err != nil {
		return nil, err
	}

	// Aggregate results into TemplateStats struct
	stats := &TemplateStats{
		TemplateID:            templateID,
		CompletedRooms:        templateStats.CompletedRooms,
		TotalRealPlayers:      templateStats.TotalRealPlayers,
		TotalBots:             templateStats.TotalBots,
		AvgRealPlayersPerRoom: templateStats.AvgRealPlayersPerRoom,
		RealPlayerWins:        winnerStats.RealPlayerWins,
		BotWins:               winnerStats.BotWins,
		TotalBoostAmount:      boostStats.TotalBoostAmount,
		AvgBoostPerPlayer:     boostStats.AvgBoostPerPlayer,
		AvgBoostPerRoom:       boostStats.AvgBoostPerRoom,
		AvgPlacesPerPlayer:    avgPlaces,
	}

	return stats, nil
}

// parseTimeFilter converts a TimeFilter into start and end timestamps
func parseTimeFilter(filter TimeFilter) (time.Time, time.Time) {
	now := time.Now()
	var startTime, endTime time.Time

	switch filter.Period {
	case "hour":
		startTime = now.Add(-1 * time.Hour)
		endTime = now
	case "day":
		startTime = now.AddDate(0, 0, -1)
		endTime = now
	case "week":
		startTime = now.AddDate(0, 0, -7)
		endTime = now
	case "month":
		startTime = now.AddDate(0, -1, 0)
		endTime = now
	case "all":
		// Use a very old date for "all time"
		startTime = time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
		endTime = now
	case "custom":
		if filter.StartTime != nil {
			startTime = *filter.StartTime
		} else {
			startTime = time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
		}
		if filter.EndTime != nil {
			endTime = *filter.EndTime
		} else {
			endTime = now
		}
	default:
		// Default to all time
		startTime = time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
		endTime = now
	}

	return startTime, endTime
}
